src/osd/PG.cc

```diff
@@ -5316,7 +5316,15 @@ void PG::on_new_interval()
     upacting_features &= osdmap->get_xinfo(*p).features;
   }
 
-  assert(osdmap->test_flag(CEPH_OSDMAP_SORTBITWISE));
+  // we don't speak non-sortbitwise, so we'd better not go
+  // active if they somehow haven't set it!
+  if (!osdmap->test_flag(CEPH_OSDMAP_SORTBITWISE) ||
+	 // but we have to be nice in case we're sortbitwise *now*
+	 // and are processing updates from before that happened
+      (osd->osd->is_active() && osdmap->get_epoch() < osd->get_boot_epoch())) {
```

Comment:
..but this condition is..weird.  osdmap is the PG's osdmap, but that may lag, so maybe we might be processing the PG interval after the osd is active but PG is still catching up?

But perhaps it's simpler to just drop this assert entirely!  We have one in OSD::activate_map():
<pre>
  if (!osdmap->test_flag(CEPH_OSDMAP_SORTBITWISE)) {
    derr << __func__ << " SORTBITWISE flag is not set" << dendl;
    ceph_abort();
  }
</pre>
so I think any check here is redundant.

Reply:
So the point here is that we only want to bail out if we're *doing work* in a PG. So we check for SORTBITWISE missing, AND that we're working in the (global) current interval. Our best local approximation for that involves making sure the OSD is active and that our map is newer than when the OSD went active...
and now I see that this condition is also backwards. No idea what the heck I was doing when I wrote this code. :/ Also changing.